### PR TITLE
tests: Cleanup subscription deployments as part of tests

### PR DESCRIPTION
**Pull request by ellismg**

### CODE REVIEW

File: cli/azd/test/functional/aspire_test.go [hunk 1]
```diff
@@ -240,6 +244,47 @@ func Test_CLI_Aspire_Deploy(t *testing.T) {
 	require.NoError(t, err)
 }
 
+// cleanupDeployments deletes all subscription level deployments tagged with `azd-env-name` equal to envName. If the session
+// indcates we are in playback mode, this function is a no-op.
+func cleanupDeployments(ctx context.Context, t *testing.T, azCLI *azdcli.CLI, session *recording.Session, envName string) {
```
Comment (ellismg) [inline cli/azd/test/functional/aspire_test.go:249]: My strategy to figure out what tests needed to call this was to look for calls to `NewCLI` where we pass in a `*Session` and arrange for these tests to call `cleanupDeployments`.

**Comment by ellismg** [comment]
Here's an internal run: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=4335287&view=results which should exercise the delete codepath (we don't get coverage from our regular PR CI runs of this, since we use playback mode).

**Comment by weikanglim** [comment]
@ellismg Just wondering about the direction here: is there value embracing the idea that automation tests try their best to simulate production behavior and avoid extra test-only deployment cleanup? For example, perhaps in the future, we may want the deployment being around as part of a production feature set.

I'm guessing that there is some ARM behavior here that is causing us to add the test-specific cleanup. My guess is that ARM is failing to keep up with deployment history purge due to the volume of deployments?

File: cli/azd/test/functional/aspire_test.go (hunk 1 above)
Comment (danieljurek) [reply cli/azd/test/functional/aspire_test.go:249]: The engsys cleanup script now also deletes sub-wide deployments. Thanks for looking for these. I couldn't find any others in my cursory searches... but my knowledge of the test suite is not as exhaustive as it was previously.

**Comment by danieljurek** [comment]
> @ellismg Thanks for doing this! Just wondering about the direction here before approving: is there value embracing the idea that automation tests try their best to simulate production behavior and avoid extra test-only deployment cleanup? For example, perhaps in the future, we may want the deployment being around as part of a production feature set.
> 
> I'm guessing that there is some ARM behavior here that is causing us to add the test-specific cleanup. My guess is that ARM is failing to keep up with deployment history purge due to the volume of deployments?

Indeed, this change addresses an issue where ARM can fall behind on the volume of subscription-level deployments being created. Later sub-wide deployments fail in that scenario. This could conceivably happen in a customer's environment (so worth examining whether it should be done long term) but has most recently been observed happening in a test resources environment which sees a high volume of such deployments.

**Comment by danieljurek** [comment]
Fixes https://github.com/Azure/azure-dev/issues/4540

**Review by danieljurek — APPROVED** [review]
LGTM.

**Review by danieljurek — CHANGES_REQUESTED** [review]
The code looks good but I'm seeing a failing test in the live test run: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=4335287&view=logs&j=3748a3ab-7269-5e9a-9cea-5ecb3b6f562c&t=9f3343b3-b8c5-51bf-d443-9aa044c5c18d&l=296 ... Perhaps the cleanup should be more resilient to deployments that aren't found? Or the deployment in `provisionRemoteStateStorage` is resource group specific?

**Review by danieljurek — APPROVED** [review]
Approved. Tests are passing.

**Pull request merged**